app/src/main/java/com/example/androidtrainingproject/products/ProductDetailsActivity.kt

```diff
@@ -1,29 +1,32 @@
 package com.example.androidtrainingproject.products
 
-import androidx.compose.foundation.BorderStroke
-import androidx.compose.foundation.Image
-import androidx.compose.foundation.border
 import androidx.compose.foundation.layout.Column
 import androidx.compose.foundation.layout.fillMaxSize
 import androidx.compose.foundation.layout.fillMaxWidth
 import androidx.compose.foundation.layout.padding
-import androidx.compose.foundation.layout.width
+import androidx.compose.foundation.shape.RoundedCornerShape
 import androidx.compose.material3.Button
 import androidx.compose.material3.ButtonDefaults
 import androidx.compose.material3.Text
 import androidx.compose.runtime.Composable
 import androidx.compose.ui.Alignment
 import androidx.compose.ui.Modifier
+import androidx.compose.ui.draw.clip
 import androidx.compose.ui.graphics.Color
-import androidx.compose.ui.res.painterResource
+import androidx.compose.ui.layout.ContentScale
 import androidx.compose.ui.text.font.FontWeight
 import androidx.compose.ui.text.style.TextAlign
 import androidx.compose.ui.unit.dp
 import androidx.compose.ui.unit.sp
-import com.example.androidtrainingproject.R
+import androidx.hilt.navigation.compose.hiltViewModel
+import coil.compose.AsyncImage
 
 @Composable
-fun ProductDetails() {
+fun ProductDetails(productId: Number) {
+    val productDetailsViewModel: ProductDetailsViewModel = hiltViewModel()
+    
+    productDetailsViewModel.getProductById(productId)
```

Comment:
You shouldn't have function invocations out in the open like this as they will trigger on every single rerender. Call this function in the `init` method of the viewmodel as the viewmodel is initialized only once.

Reply:
Ok, I moved it to the viewmodel's init, but now I'm not sure how it will know what ID to call unless I hardcode it inside it... not sure if there is a way to pass it from the Activity somehow

Reply:
https://stackoverflow.com/questions/67982230/jetpack-compose-pass-parameter-to-viewmodel - second answer with the `SavedStateHandle` example. It will become even a bit easier with the navigation library that we're going to show you on the next lecture but a `SavedStateHandle` will still be used.

Reply:
Seems like I followed the code 1:1 but I still get some sort of error... I guess this will be left over for the next lecture

"Navigation destination that matches request NavDeepLinkRequest{ uri=android-app://androidx.navigation/ProductDetails } cannot be found in the navigation graph NavGraph(0x0) startDestination={Destination(0xa0a5cfdc) route=Login}"